processor/deltatocumulativeprocessor/internal/delta/delta.go

```diff
@@ -0,0 +1,83 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package delta // import "github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/delta"
+
+import (
+	"fmt"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/data"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/processor/deltatocumulativeprocessor/internal/streams"
+)
+
+func construct[D data.Point[D]]() streams.Aggregator[D] {
+	acc := &Accumulator[D]{dps: make(map[streams.Ident]D)}
+	return &Lock[D]{next: acc}
+}
+
+func Numbers() streams.Aggregator[data.Number] {
+	return construct[data.Number]()
+}
+
+func Histograms() streams.Aggregator[data.Histogram] {
+	return construct[data.Histogram]()
+}
+
+var _ streams.Aggregator[data.Number] = (*Accumulator[data.Number])(nil)
+
+type Accumulator[D data.Point[D]] struct {
+	dps map[streams.Ident]D
+}
+
+// Aggregate implements delta-to-cumulative aggregation as per spec:
+// https://opentelemetry.io/docs/specs/otel/metrics/data-model/#sums-delta-to-cumulative
+func (a *Accumulator[D]) Aggregate(id streams.Ident, dp D) (D, error) {
+	// make the accumulator to start with the current sample, discarding any
+	// earlier data. return after use
+	reset := func() (D, error) {
+		a.dps[id] = dp.Clone()
+		return a.dps[id], nil
+	}
+
+	aggr, ok := a.dps[id]
+
+	// new series: reset
+	if !ok {
+		return reset()
+	}
+	// belongs to older series: drop
+	if dp.StartTimestamp() < aggr.StartTimestamp() {
+		return aggr, ErrOlderStart{Start: aggr.StartTimestamp(), Sample: dp.StartTimestamp()}
+	}
+	// belongs to later series: reset
+	if dp.StartTimestamp() > aggr.StartTimestamp() {
+		return reset()
+	}
+	// out of order: drop
+	if dp.Timestamp() <= aggr.Timestamp() {
+		return aggr, ErrOutOfOrder{Last: aggr.Timestamp(), Sample: dp.Timestamp()}
+	}
+
+	a.dps[id] = aggr.Add(dp)
+	return a.dps[id], nil
+}
+
+type ErrOlderStart struct {
+	Start  pcommon.Timestamp
+	Sample pcommon.Timestamp
+}
+
+func (e ErrOlderStart) Error() string {
+	return fmt.Sprintf("dropped sample with st=%s, because series only starts at st=%s", e.Sample, e.Start)
```

Comment:
As someone reading this message, I might not know what `st` is. This is a timestamp, so, it might be good to spell it out: "dropped sample with timestamp=%s". Bonus points if you can give an action to the user: "check if XYZ is correct".

Reply:
what would be an action item we could give the user when samples of the same series identity have different start times? check for two processes emitting the exact same series?

Reply:
Yes, telling them about possible causes is a great start.